presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -466,6 +486,61 @@ private RelationPlan createTableWriterPlan(
         return new RelationPlan(commitNode, analysis.getRootScope(), commitNode.getOutputSymbols());
     }
 
+    /*
+    According to the standard, for the purpose of store assignment (INSERT),
+    no non-space characters of a character string, and no non-zero octets
+    of a binary string must be lost when the inserted value is truncated to
+    fit in the target column type.
+    The following method returns a cast from source type to target type
+    with a guarantee of no illegal truncation.
+    TODO Once BINARY and parametric VARBINARY types are supported, they should be handled here.
+    TODO This workaround is insufficient to handle structural types
+     */
+    private Expression noTruncationCast(Expression fromExpression, Type fromType, Type toType)
+    {
+        if (fromType instanceof UnknownType || (!(toType instanceof VarcharType) && !(toType instanceof CharType))) {
+            return new Cast(fromExpression, toSqlType(toType));
+        }
+        int targetLength;
+        if (toType instanceof VarcharType) {
+            if (((VarcharType) toType).isUnbounded()) {
+                return new Cast(fromExpression, toSqlType(toType));
+            }
+            targetLength = ((VarcharType) toType).getBoundedLength();
+        }
+        else {
+            targetLength = ((CharType) toType).getLength();
+        }
+
+        checkState(fromType instanceof VarcharType || fromType instanceof CharType, "inserting non-character value to column of character type");
+        Type regexType = regexLibrary.name().equals("JONI") ? JONI_REGEXP : RE2J_REGEXP;
+        ResolvedFunction regexpReplace = metadata.resolveFunction(QualifiedName.of("regexp_replace"), fromTypes(VARCHAR, regexType));
+        ResolvedFunction length = metadata.resolveFunction(QualifiedName.of("length"), fromTypes(VARCHAR));
+        ResolvedFunction fail = metadata.resolveFunction(QualifiedName.of("fail"), fromTypes(VARCHAR));
+
+        return new IfExpression(
+                // check if the trimmed value fits in the target type
+                new ComparisonExpression(
+                        GREATER_THAN_OR_EQUAL,
+                        new GenericLiteral("BIGINT", Integer.toString(targetLength)),
+                        // length of the trimmed value
+                        new FunctionCall(
+                                length.toQualifiedName(),
+                                ImmutableList.of(
+                                        // remove trailing spaces
+                                        new FunctionCall(
+                                                regexpReplace.toQualifiedName(),
```

Comment:
Have you considered using the `rtrim` function, instead?

Reply:
According to the documentation, the `rtrim` function removes all whitespace characters. And the SQL Standard only allows to truncate spaces during insert.

Reply:
But `rtrim` would remove the trailing whitespace. Why can't we go for rule based optimizer here , which would do the casting.